common/backpressure/moving_window_avg.go

```diff
@@ -0,0 +1,80 @@
+package backpressure
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+type (
+	TimestampedData struct {
+		Value     int64
+		Timestamp time.Time
+	}
+
+	MovingWindowAvg interface {
+		Add(val int64)
+		Average() float64
+	}
+
+	MovingAvgWindowImpl struct {
+		windowSize    time.Duration
+		buffer        chan TimestampedData
```

Comment:
I'm a little wary of using a channel as a buffer since that's not really what they're designed for. Do you know if the implementation uses a ring buffer or something else constant time?

Overall, it's possible to implement a moving window without any channels and extra goroutines, and personally I think it might be easier to understand. Why do you think this way is better?

Reply:
I believe the underlying buffer implementation is a linked list, so yes constant time. The primary reason I picked a buffered channel over golang's builtin `ring` type was to reduce lock contention. Builtin `ring` buffers are not thread safe and I expect these methods to be used very heavily from concurrent calls and so I would like to keep the performance impact on persistence calls to an absolute minimum. And even with a `ring` I think we would still need a separate goroutine to expire elements based on a time value; putting that logic in the Add method could also cause a performance hit.

I did look for other examples of ways to do this, but maybe I missed something because I am still fairly new to golang. Did you have any examples you could send me if you think there is a better way?

Reply:
yeah I would also suggest write some benchmark test for this component and see how to performs. If it can easily handle ~5k (the per pod persistence rps limit in our deployment is lower than that number) updates per second then I don't have any concern for the implementation.